File: server/src/protocol.ts
export const MessageType = {
  MOVE: 0x01,
  TOUCH_DOWN: 0x02,
  TOUCH_UP: 0x03,
  TAP: 0x04,
} as const;

export type MessageType = (typeof MessageType)[keyof typeof MessageType];

export interface Message {
  type: MessageType;
  x: number;
  y: number;
}

export function decode(data: Buffer): Message | null {
  if (data.length !== 5) return null;

  const type = data[0] as MessageType;
  if (type < 0x01 || type > 0x04) return null;

  const x = (data[1] << 8) | data[2];
  const y = (data[3] << 8) | data[4];

  return { type, x, y };
}

File: server-rust/src/protocol.rs
#[derive(Clone, Copy, Debug)]
pub enum MessageType {
  Move = 0x01,
  TouchDown = 0x02,
  TouchUp = 0x03,
  Tap = 0x04,
}

#[derive(Clone, Copy, Debug)]
pub struct Message {
  pub kind: MessageType,
  pub x: i32,
  pub y: i32,
}

pub fn decode(data: &[u8]) -> Option<Message> {
  if data.len() != 5 {
    return None;
  }

  let kind = match data[0] {
    0x01 => MessageType::Move,
    0x02 => MessageType::TouchDown,
    0x03 => MessageType::TouchUp,
    0x04 => MessageType::Tap,
    _ => return None,
  };

  let x = u16::from_be_bytes([data[1], data[2]]) as i32;
  let y = u16::from_be_bytes([data[3], data[4]]) as i32;

  Some(Message { kind, x, y })
}
